### ISSUE TITLE
[BUG] Volume stuck in Detaching when detached during rebuild / snapshot coalescing (EXT4)

### ISSUE THREAD

**Issue by joshimoo** [longhorn/longhorn#2640]
This happens since the FIEMAP call gives control to the kernel so the replica process cannot be stopped.
While the kernel is retrieving the extents. On EXT4 and kernel versions below < `5.8.0-53-generic`

TODO: narrow this down to the correct patched version + changelog
`5.4.0-73-generic` = slow, `5.8.0-53-generic` = fast

On the iterative implementation added in #2507 #1948
This will be less of an issue since the delay for return to the process is time(1024 extent retrieval)
On a test node with 10GB of 4K data / 4K holes, this retrieval took around 3 minutes.
As one can see it will take hours for the replica process to be able to terminate in the previous implementation.
During that time the volume will be stuck in detaching state.

TODO: consider adding kernel version / filesystem detection to determine the extent count per retrieval, since time spend is proportional to the extents retrieved.

CONTEXT:
The instance manager will be spamming this message multiple times into the instance-manager log.
When this case is hit.
```
[longhorn-instance-manager] time="2021-05-28T19:38:47Z" level=debug msg="Process Manager: wait for process rebuild-perf-test-old-r-869abec8 to shutdown before unregistering process"
```

**Comment by yasker** [comment]
I thought FIEMAP issue only happens with bigger volumes with large amount of extents. Do we know in which case it can take hours? Also, do we have any existing issues with similar error log? (Well, sounds like we need an elastic search cluster for support bundle).

**Comment by joshimoo** [comment]
With the iterative processing for rebuilds the process can be cancelled.
So this should be mitigated on older kernels after the enhancement.
It might still take up to 3 minutes to kill the process but it will terminate once the 1024 extent retrieval has returned from the kernel.

**Comment by joshimoo** [comment]
For additional setup/testing instructions you can check the comments in that issue: 
https://github.com/longhorn/longhorn/issues/2507#issuecomment-857184562

The test is quite simple here, try detaching a volume during rebuild on an older kernel.
Since the process is stuck in the kernel till it returns from the extent processing the replica process will not stop.
And the following will not become available. The replica process will continue consuming resources on the node.
Even if it becomes defunct for whatever reason.

After the enhancement the retrieval of 1024 extents should only take a maximum of 3 minutes so the process should be able to be killed which will allow the volume to enter the detached state.

**Comment by khushboo-rancher** [comment]
Verified with Longhorn-v1.1.2-rc1 and Longhorn-master-head `07/28/2021`

Validation - **Pass**

Validated with Ubuntu 20.04, kernel version 5.4, Data size 10 Gi 4K data / 4K holes.
With v.1.1.1, I was able to reproduce the problem which is fixed on Longhorn v1.1.2-rc1 and master.

**Steps**
1. Create a volume (3 replicas) and attach it to a node.
2. Write Data size 10 Gi 4K data / 4K holes.
3. Delete a replica, this will trigger rebuilding of a replica.
4. While replica is in progress.
5. Detach the volume from the node.
